### PR TITLE
OpenAPI oneOf support

### PR DESCRIPTION
**Pull request by CharlesDuboisSAP**
## Context

AI/ai-sdk-java-backlog#112

oneOf / anyOf / allOf with multiple possible options should form an interface or a class that is able to parse those multiple options
[openAPI templates were updated to 7.8.0](https://github.com/OpenAPITools/openapi-generator/tree/v7.8.0/modules/openapi-generator/src/main/resources/Java)

### Feature scope:
 
- [x] oneOf support

## Definition of Done

- [x] Functionality scope stated & covered
- [x] Tests cover the scope above
- [ ] ~Error handling created / updated & covered by the tests above~
- [x] Documentation updated [Added OpenAPI generator options generateApis and generateModels](https://github.com/SAP/cloud-sdk/pull/1953#top)
- [x] Release notes updated

### CODE REVIEW

File: datamodel/openapi/openapi-api-sample/src/main/java/com/sap/cloud/sdk/datamodel/openapi/sample/model/Order.java [hunk 1]
```diff
@@ -154,7 +154,7 @@ public Order totalPrice( @Nullable final Float totalPrice )
      *
      * @return totalPrice The totalPrice of this {@link Order} instance.
      */
-    @Nonnull
+    @Nullable
```
Comment (newtork) [inline datamodel/openapi/openapi-api-sample/src/main/java/com/sap/cloud/sdk/datamodel/openapi/sample/model/Order.java:157]: (Question)

Is this... a drive-by change?

Comment (CharlesDuboisSAP) [reply datamodel/openapi/openapi-api-sample/src/main/java/com/sap/cloud/sdk/datamodel/openapi/sample/model/Order.java:157]: Yes, here totalPrice is not a required field

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/oneof/input/sodastore.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        '200':
+          description: A list of sodas
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/Soda'
+
+components:
```
Comment (newtork) [inline datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/oneof/input/sodastore.yaml:22]: (Major)

I like the technical test setup. But semantically we can still improve a little:

* Fanta: has caffeine? no bubbles?
* **Implicitly** your test spec assumes strange values:
  * `Fanta.caffeine = "Fanta"`
  * `Cola.caffeine = "Cola"`

---

While the feature is behind a toggle, I'm fine with not having it 100% coverage tested.
* What happens with `anyof`?
* What happens with missing discriminator?
* What happens with discriminator + value mapping?

File: datamodel/openapi/openapi-api-sample/src/main/java/com/sap/cloud/sdk/datamodel/openapi/sample/model/Order.java (hunk 1 above)
Comment (newtork) [reply datamodel/openapi/openapi-api-sample/src/main/java/com/sap/cloud/sdk/datamodel/openapi/sample/model/Order.java:157]: Can you make this change a separate PR with release notes then? Looks like a nice improvement to me.

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/input-spec-with-anyof-oneof/output/com/sap/cloud/sdk/services/anyofoneof/model/RootObjectQuestionsInner.java [hunk 3]
```diff
@@ -43,10 +42,10 @@
 /**
  * RootObjectQuestionsInner
  */
+@JsonTypeName("RootObject_questions_inner")
```
[the remark this answers is not in the document]
Comment (CharlesDuboisSAP) [reply datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/input-spec-with-anyof-oneof/output/com/sap/cloud/sdk/services/anyofoneof/model/RootObjectQuestionsInner.java]: All tests in AI SDK are green, I can still remove it if you want.

[the remark this answers is not in the document]
Comment (CharlesDuboisSAP) [reply datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/input-spec-with-anyof-oneof/output/com/sap/cloud/sdk/services/anyofoneof/model/RootObjectQuestionsInner.java]: Removed

File: datamodel/openapi/openapi-generator-maven-plugin/src/main/java/com/sap/cloud/sdk/datamodel/openapi/generator/DataModelGeneratorMojo.java [hunk 4]
```diff
@@ -118,6 +118,18 @@ public class DataModelGeneratorMojo extends AbstractMojo
...
+     */
+    @Parameter( property = "openapi.generate.generateModels", defaultValue = "true" )
+    private boolean generateModels;
+
+    /**
+     * Generate API classes (client classes). Default is true.
+     */
+    @Parameter( property = "openapi.generate.generateApis", defaultValue = "true" )
+    private boolean generateApis;
+
```
Comment (newtork) [inline datamodel/openapi/openapi-generator-maven-plugin/src/main/java/com/sap/cloud/sdk/datamodel/openapi/generator/DataModelGeneratorMojo.java:132]: (Question)

Why not have this as separate PR? It's unrelated to oneof / anyof.

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/input/sodastore.yaml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        - $ref: '#/components/schemas/Fanta'
+    AnyOf:
+      oneOf:
+        - $ref: '#/components/schemas/Cola'
+        - $ref: '#/components/schemas/Fanta'
+
+    AllOf:
+      oneOf:
+        - $ref: '#/components/schemas/Cola'
+        - $ref: '#/components/schemas/Fanta'
```
Comment (newtork) [inline datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/input/sodastore.yaml:54]: (Minor)

Inconsistent component schema names.

File: datamodel/openapi/openapi-generator/src/test/java/com/sap/cloud/sdk/datamodel/openapi/generator/DataModelGeneratorIntegrationTest.java [hunk 6]
```diff
@@ -113,15 +120,27 @@ private enum TestCase
             "test",
             ApiMaturity.RELEASED,
             true,
-            4,
-            Map.of("useOneOfInterfaces", "true")),;
+            true,
+            6,
+            Map.of("useOneOfInterfaces", "true")),
+        GENERATE_APIS(
```
Comment (newtork) [inline datamodel/openapi/openapi-generator/src/test/java/com/sap/cloud/sdk/datamodel/openapi/generator/DataModelGeneratorIntegrationTest.java:126]: (Comment)

Not really sure, whether it's good to mix testing scopes of generate-api and anyof/allof/oneof.
But I'm not objecting at this point.

File: release_notes.md [hunk 7]
```diff
@@ -13,6 +13,12 @@
 ### ✨ New Functionality
 
 - Add support for `TypeDefinition` entries in OData V4 EDMX files.
+- OpenAPI generator improvements:
+  - Add `oneOf`, `anyOf` and `allOf` support for OpenAPI generation.
```
Comment (newtork) [inline release_notes.md:17]: (Major)

I'll check the PR in AI SDK, but still it would be good to have a proper unit test, checking whether serialization and deserialization works. We are not even checking whether compilation works, right?

File: datamodel/openapi/openapi-generator-maven-plugin/src/main/java/com/sap/cloud/sdk/datamodel/openapi/generator/DataModelGeneratorMojo.java (hunk 4 above)
Comment (CharlesDuboisSAP) [reply datamodel/openapi/openapi-generator-maven-plugin/src/main/java/com/sap/cloud/sdk/datamodel/openapi/generator/DataModelGeneratorMojo.java:132]: to save time

File: release_notes.md (hunk 7 above)
Comment (CharlesDuboisSAP) [reply release_notes.md:17]: Checking for compilation in the sample

Comment (CharlesDuboisSAP) [reply release_notes.md]: Also this is a hidden feature in the additional parameters. It should receive the same tests as your pojoBuilder

Comment (newtork) [reply release_notes.md]: > Checking for compilation in the sample

Please link the sample that is being compiled.

> this is a hidden feature in the additional parameters

I think this is incorrect, ["enableOneOfAnyOfGeneration" is a well documented](https://sap.github.io/cloud-sdk/docs/java/features/rest/generate-rest-client) and [used feature](https://github.com/SAP/cloud-sdk-java/issues/312). It's not considered _hidden_ compared to the _pojoBuilder_.

If we claim support for oneof, anyof and allof - then we should have them tested for deserialization/serialization. If I see correctly the `openapi-api-sample` module only tests for `allof` currently.

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/output/test/OneOf.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * OneOf
+ */
+// CHECKSTYLE:OFF
+public class OneOf
```
Comment (MatKuhr) [inline datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/output/test/OneOf.java:37]: why is this a class and not an interface?

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/input/sodastore.yaml [hunk 9]
```diff
@@ -0,0 +1,63 @@
+openapi: 3.0.0
```
Comment (MatKuhr) [inline datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/input/sodastore.yaml:1]: validating this file in the editor reveals 3 violations

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/output/test/OneOfWithDiscriminatorAndMapping.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+import javax.annotation.Nullable;
+
+/**
+ * OneOfWithDiscriminatorAndMapping
+ */
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "sodaType", visible = true)
+@JsonSubTypes({
+  @JsonSubTypes.Type(value = Cola.class, name = "Cola"),
+  @JsonSubTypes.Type(value = Fanta.class, name = "Fanta"),
```
Comment (MatKuhr) [inline datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/output/test/OneOfWithDiscriminatorAndMapping.java:43]: How come the discriminator mapping doesn't seem to have any effect? (comparing to the other object that doesn't have the mapping)

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/input-spec-with-anyof-oneof/output/com/sap/cloud/sdk/services/anyofoneof/model/RootObjectQuestionsInner.java [hunk 11]
```diff
@@ -75,7 +73,8 @@ public class RootObjectQuestionsInner
    * Get id
    * @return id  The id of this {@link RootObjectQuestionsInner} instance.
```
Comment (MatKuhr) [inline datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/input-spec-with-anyof-oneof/output/com/sap/cloud/sdk/services/anyofoneof/model/RootObjectQuestionsInner.java:74]: shouldn't this change to an interface?

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/oneof/input/sodastore.yaml [hunk 12]
```diff
@@ -0,0 +1,63 @@
+openapi: 3.0.0
+info:
```
Comment (MatKuhr) [inline datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/oneof/input/sodastore.yaml:2]: why do we have this spec file twice now? Also, why not re-use the existing `DataModelGeneratorIntegrationTest/input-spec-with-anyof-oneof/input/AggregatorNestedSchemaChild.json` ?

I know we already have a rather bloated test setup here, with much duplication, but adding more to it will cause PR diffs to get larger and larger..

**Review by MatKuhr — DISMISSED** [review]
some comments, most importantly though I think we need a compilation and a runtime check, i.e. sample code and test assertions on (de-) serialization

File: release_notes.md (hunk 7 above)
Comment (CharlesDuboisSAP) [reply release_notes.md]: I added a deserialization test in the sample app

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/input/sodastore.yaml (hunk 9 above)
Comment (CharlesDuboisSAP) [reply datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/input/sodastore.yaml:1]: fixed

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/output/test/OneOf.java (hunk 8 above)
Comment (CharlesDuboisSAP) [reply datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/output/test/OneOf.java:44]: This is testing generateApis false not oneOf stuff

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/output/test/OneOfWithDiscriminatorAndMapping.java (hunk 10 above)
Comment (CharlesDuboisSAP) [reply datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/generate-apis/output/test/OneOfWithDiscriminatorAndMapping.java:50]: Because it can figure out the mapping automatically

File: datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/input-spec-with-anyof-oneof/output/com/sap/cloud/sdk/services/anyofoneof/model/RootObjectQuestionsInner.java (hunk 11 above)
Comment (CharlesDuboisSAP) [reply datamodel/openapi/openapi-generator/src/test/resources/DataModelGeneratorIntegrationTest/input-spec-with-anyof-oneof/output/com/sap/cloud/sdk/services/anyofoneof/model/RootObjectQuestionsInner.java]: Nope, useOneOfInterfaces is disabled in this test

File: datamodel/openapi/openapi-api-sample/pom.xml [hunk 13]
```diff
@@ -108,7 +113,7 @@
 					</execution>
 				</executions>
 				<configuration>
-					<inputSpec>${project.basedir}/src/main/resources/sodastore.json</inputSpec>
+					<inputSpec>${project.basedir}/src/main/resources/sodastore.yaml</inputSpec>
```
Comment (newtork) [inline datamodel/openapi/openapi-api-sample/pom.xml:116]: (Question)

Why did you change from json to yaml?
Only for flavor, or does anyof/oneof not work for json?

Comment (CharlesDuboisSAP) [reply datamodel/openapi/openapi-api-sample/pom.xml:116]: Flavor, the file is shorter

File: datamodel/openapi/openapi-api-sample/src/test/java/com/sap/cloud/sdk/datamodel/openapi/sample/api/OneOfDeserializationTest.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    void allOf()
+        throws JsonProcessingException
+    {
+        AllOf allOfCola = newDefaultObjectMapper().readValue(cola, AllOf.class);
+        assertThat(allOfCola.getSodaType()).isEqualTo("Cola");
+        assertThat(allOfCola.isCaffeine()).isTrue();
+
+        AllOf allOfFanta = newDefaultObjectMapper().readValue(fanta, AllOf.class);
+        assertThat(allOfFanta.getSodaType()).isEqualTo("Fanta");
+        assertThat(allOfFanta.getColor()).isEqualTo("orange");
```
Comment (newtork) [inline datamodel/openapi/openapi-api-sample/src/test/java/com/sap/cloud/sdk/datamodel/openapi/sample/api/OneOfDeserializationTest.java:100]:
```suggestion
        assertThat(allOfFanta.getColor()).isEqualTo("orange");
        assertThat(allOfFanta.isCaffeine()).isNull();
```

**Review by newtork — APPROVED** [review]

**Review by MatKuhr — APPROVED** [review]
Otherwise lgtm
